File: 0131-palindrome-partitioning/0131-palindrome-partitioning.java
class Solution {
    List<List<String>> res;
    String s;
    public List<List<String>> partition(String s) {
        this.res = new ArrayList<>();
        this.s = s;
        backtrack(new LinkedList<>(), 0);
        return res;
    }
    
    private void backtrack(LinkedList<String> comb, int start) {
        if (start >= s.length()) {
            res.add(new LinkedList<>(comb));
        }
        for (int i = start; i < s.length(); i++) {
            if (isPalindrome(start, i)) {
                comb.add(s.substring(start, i+1));
                backtrack(comb, i+1);
                comb.removeLast();
            }
        }
    }
    
    private boolean isPalindrome(int start, int end) {
        while (start < end) {
            if (s.charAt(start++) != s.charAt(end--)) {
                return false;
            }
        }
        return true;
    }
}

File: 0131-palindrome-partitioning/0131-palindrome-partitioning.cpp
class Solution {
public: 
    vector<vector<string>> partition(string s) {
        this->s = s;
        vector<string> comb;
        backtrack(comb, 0);
        return res;
    }
    
private:
    vector<vector<string>> res;
    string s;
    void backtrack(vector<string> comb, int start) {
        if (start >= s.size()) {
            res.push_back(comb);
        }
        for (int i = start; i < s.size(); i++) {
            if (isPalindrome(start, i)) {
                comb.push_back(s.substr(start, i-start+1));
                backtrack(comb, i+1);
                comb.pop_back();
            }
        }
    }
    
    bool isPalindrome(int start, int end) {
        while (start < end) {
            if (s[start++] != s[end--]) {
                return false;
            }
        }
        return true;
    }
};